include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -1068,50 +1092,349 @@ namespace Differentiation
-                                 const unsigned int /*direction*/)
+
+      /**
+       * Type for real numbers.
+       */
+      using real_type =
+        typename dealii::numbers::NumberTraits<scalar_type>::real_type;
+
+
+      /**
+       * Type for complex numbers.
+       */
+      using complex_type = std::complex<real_type>;
+
+
+      /**
+       * The actual auto-differentiable number type.
+       */
+      using ad_type = typename std::
+        conditional<is_real_valued, real_type, complex_type>::type;
+
+      /**
+       * The actual auto-differentiable number directional derivative type.
+       */
+      using derivative_type = typename std::
+        conditional<is_real_valued, real_type, complex_type>::type;
+
+
+      /**
+       * Extract the value of an auto-differentiable number.
+       */
+      static scalar_type
+      get_scalar_value(const ad_type &x)
+      {
+        return x;
+      }
+
+
+      /**
+       * Extract the derivative value of an auto-differentiable number.
+       */
+      static derivative_type
+      get_directional_derivative(const ad_type &, const unsigned int)
+      {
+        Assert(
+          false,
+          ExcMessage(
+            "Floating point/arithmetic numbers have no directional derivatives."));
+        return derivative_type();
+      }
+
+
+      /**
+       * Extract the number of directional derivatives value tracked by
+       * an auto-differentiable number.
+       */
+      static unsigned int
+      n_directional_derivatives(const ad_type &)
       {
-        // If the AD drivers are correctly implemented then we should not get
-        // here. This is essentially a dummy for when the ADNumberTypeCode for
-        // the original AD number (from which one is getting a derivative >= 2)
-        // is one that specified ADOL-C taped and tapeless numbers, or a
-        // non-nested Sacado number.
-        AssertThrow(
+        Assert(
           false,
           ExcMessage(
-            "Floating point numbers have no directional derivatives."));
-        return 0.0;
+            "Floating point/arithmetic numbers have no directional derivatives."));
+        return 0;
       }
+
+
+      static_assert(std::is_same<ad_type, ScalarType>::value,
+                    "Incorrect template type selected for ad_type");
+
+      static_assert((is_complex_valued == true ?
+                       boost::is_complex<ad_type>::value :
+                       true),
+                    "Expected a complex ad_type");
     };
 
+#  ifdef __clang__
+
+    template <typename ScalarType>
+    const bool NumberTraits<
+      ScalarType,
+      NumberTypes::none,
+      typename std::enable_if<
+        std::is_floating_point<ScalarType>::value ||
+        (boost::is_complex<ScalarType>::value &&
+         std::is_floating_point<typename internal::RemoveComplexWrapper<
+           ScalarType>::type>::value)>::type>::is_taped = false;
```

Comment:
Good god. Who can read this? ;-)

Reply:
Nobody... Its so ugly to look at that I had to write it with my eyes closed >-|

Reply:
:-)